Support wasmJs target

IlyaGulya opened the pull request:
What's impossible for now:
1) Commonization of `WebWorkerDriver` can't be done without support for Kotlin arrays interop with `wasmJs`: https://youtrack.jetbrains.com/issue/KT-57125/Add-interop-with-JavaScript-arrays-to-Kotlin-Wasm

Changes:
1) Enable `wasmJs` target in `MultiplatformConventions` and `:driver-test`.
2) Migrate to use kotlin multiplatform default hierarchy template.
3) Add `jsCommon` source set containing common code for `js` and `wasmJs` targets.
4) Added `jsNativeCommon` source set to `:runtime` module containing common code for `native`, `js`, and `wasmJs` targets.
5) Enable `watchosDeviceArm64` and `linuxArm64` because of failing tests.
6) Support `WebWorkerDriver` for `wasmJs` target. Fix all tests to be passing.

dellisd commented:
Is the conversion to `.gradle.kts` required to enable the `wasmJs` target? If not, it would be best to leave that conversion for a different PR since this one has a lot of files that are just moved and it's more difficult to review the actual changes here.

IlyaGulya commented:
Sure, I will revert this change

IlyaGulya commented:
@dellisd Hi!
I've rebased on the latest master.
Here are changes since your comment:
1. Reverted the migration to `.gradle.kts`.
2. Made `web-worker-driver` tests for `js` and `wasmJs` common.
3. Made `createWebWorkerDriver` helper function which allows creating the default `WebWorkerDriver` in the `jsCommon` sourceSet.
4. Upgraded Kotlin to `2.0.0-Beta3` and KSP to `2.0.0-Beta3-1.0.17`

IlyaGulya commented:
@dellisd Hi!
I've rebased on latest master.
Here're changes since my previous comment:

1. Upgraded Kotlin to `2.0.0-Beta4` and KSP to `2.0.0-Beta4-1.0.17`
2. Upgraded `kotlinx.coroutines` to `1.8.0` with `wasmJs` support
3. Temporarily added Sonatype repository containing Turbine `1.1.0-SNAPSHOT`

IlyaGulya commented:
I'm not sure why, but instrumentation tests are failing on CI.
They're passing locally on my M1 Max.
https://github.com/IlyaGulya/sqldelight/actions/runs/7957378205/job/21719986393

artemyto commented:
Turbine 1.1.0 with wasmJs support is out

IlyaGulya commented:
Replaced Turbine snapshot with `1.1.0`

IlyaGulya commented:
Currently trying to fix tests 🙂

dellisd commented on drivers/web-worker-driver/build.gradle:
```diff
@@ -5,39 +5,43 @@ plugins {
   }
+  applyDefaultHierarchyTemplate {
+    it.common {
+      it.group("jsCommon") {
```
Is there any need for this to be `jsCommon` instead of leaving it as the default `common`?

IlyaGulya replied:
Yeah, you are right, my bad, I'll remove this group.

dellisd commented on drivers/web-worker-driver/src/jsCommonMain/kotlin/app/cash/sqldelight/driver/worker/CreateWebWorkerDriver.kt:
```diff
@@ -0,0 +1,5 @@
+
+import app.cash.sqldelight.db.SqlDriver
+
+expect fun createWebWorkerDriver(): SqlDriver
```
The web worker driver is designed to support more than just one worker implementation, so this function should be able to accept a `Worker` parameter to allow a consumer to define which worker should be loaded.

IlyaGulya replied:
It is still possible to create WebWorkerDriver with custom `Worker`.
I think we could just rename this method to `createDefaultWebWorkerDriver()` instead.

dellisd commented on drivers/web-worker-driver/src/wasmJsMain/kotlin/app/cash/sqldelight/driver/worker/WebWorkerDriver.kt:
```diff
@@ -0,0 +1,240 @@
+ * @property worker The Worker running a SQL implementation that this driver communicates with.
+ * @see [WebWorkerDriver.fromScriptUrl]
+ */
+class WebWorkerDriver(private val worker: Worker) : SqlDriver {
```
Which elements of this class (and its `jsMain` counterpart) can't be lifted into the `commonMain` implementation? Ideally I think we would have one common implementation of the worker driver, and then `expect`/`actual` the web API bits (i.e. `Worker`) as needed.

IlyaGulya replied:
TBH, most of it =)
I have tried to make it as common as possible, but found out that currently there's no way to do it properly.

Most of JS APIs are not common between `js` and `wasmJs`.
Current `wasmJs` interop is quite limited.
It does not support: `Array, dynamic` in external types.
This means that we cannot share `WorkerRequest`, `WorkerResult`, `WorkerResponse`.
Also, we would need to expect/actual quite big portion of `Worker` API with custom wrappers all over the place.

I will try some more ways to commonise as much as possible and show you in a couple of days, maybe we will find a better solution.

IlyaGulya replied:
@dellisd I've tried to commonize `WebWorkerDriver` as much as possible. Please take a look 🙂  https://github.com/cashapp/sqldelight/pull/4965/commits/ece537a35558278801051c2b74c8b9dc91b0d433

dellisd reviewed (commented):
Overall this looks good, but there are some changes to the public API that I think we should make and I'm a little unclear on why certain parts of the Wasm implementation can't be shared with the Js implementation.

IlyaGulya commented on drivers/driver-test/build.gradle:
```diff
@@ -23,9 +26,11 @@ kotlin {
   watchosX64()
   watchosArm32()
   watchosArm64()
+  watchosDeviceArm64()
```
Done!
I don't really remember why I've added them. AFAIK, because of some tests failing, but I can't reproduce it now.

hfhbd commented:
Do you need Kotlin 2.0 for this change too? I bumped it in this PR #5217, but I will break my PR into smaller ones.

IlyaGulya commented:
> Do you need Kotlin 2.0 for this change too? I bumped it in this PR #5217, but I will break my PR into smaller ones.

Not really.
But I will be happy to rebase on 2.0.0 once it's merged 🙂

shriharsha-srb commented:
Is there any reason why this is not being merged/released?

tamimattafi commented:
Any updates on this?

dpnkrg commented:
Any update on the same.

sunildhiman90 commented:
Any update on this please?

petrstetka commented:
I am also interested in wasm support, it would be very nice to have it ready.

ahna92 commented:
can we get an alpha (early access) version with this change to try it out ? 🙏

ghasemdev commented:
I was developing a multiplatform library and needed to implement a database. After adding SQLDelight, I realized that the plugin causes errors in the wasmjs module, essentially forcing me to remove this module because there's no way to ignore the error and use SQLDelight on other platforms while handling storage differently in wasm. Currently, libraries like Koin, Ktor, ViewModel, and Lifecycle support wasm, and the lack of SQLDelight support is noticeable.

bakjoul commented:
Hello. What's the current state of using sqldelight with wasmjs today ?
I see that i still get an errors with warmJs.

dpnkrg commented:
any update on the same ?

dpnkrg commented:
Still any update on the same

ahna92 commented:
Hi 
Until this MR is merged , 
I've created a no operation wasm target version `2.1.0-wasm-no-op.2`

Need to make some changes while adding the lib

update `settings.gradle.kts`
```diff
pluginManagement {
    repositories {
        mavenLocal()
        google()
        gradlePluginPortal()
        mavenCentral()
+        maven { url = uri("https://s01.oss.sonatype.org/content/groups/staging") } // <---- add this
    }
}

dependencyResolutionManagement {
    repositories {
        mavenLocal()
        google()
        mavenCentral()
+        maven { url = uri("https://s01.oss.sonatype.org/content/groups/staging") } // <---- add this
    }
}
```

then update only plugin and library group, `app.cash.sqldelight` -> `io.telereso.kmp.sqldelight`

```diff
-id("app.cash.sqldelight") version "2.0.2"
+id("io.telereso.kmp.sqldelight") version "2.1.0-wasm-no-op.2"
```

```diff
-implementation("app.cash.sqldelight:sqlite-driver:2.0.2")
+implementation("app.cash.sqldelight:sqlite-driver:2.1.0-wasm-no-op.2")
```

**No need to update src classes imports**
___

My use case that I had 9 project with db I needed to expose to web , 
I've tested 1 so far it worked 
I had to change the logic to not store/fetch from db in browser target, but rather use network directly 


_I'll try to publish this MR too to see if it works or not , but need some time_

GGsrvg commented:
@dellisd any update?

IlyaGulya commented:
Upgraded to latest master, fixed all errors.

dellisd commented:
I appreciate the huge amount of enthusiasm around this PR, and I want to say that I'm also very excited about the prospect of using SQLDelight in Wasm projects.

I also really appreciate the contributions being made here, but this PR is very difficult to review as-is because of its size and because of all of the changes that aren't directly related to enabling Wasm that are being pulled into here.

My recommendation would be to break this PR up into a few smaller PRs that are more focused and will be easier/quicker to review:

1. A PR that enables the `wasmJs()` target in the runtime and extensions
2. A PR that adds Wasm support for the web worker driver
3. Additional PRs to make other gradle config/code cleanliness changes after the fact

Thank you!

ahna92 commented:
@dellisd Agree, there are lots of changes to just accept , 

but until it's broken down,  can we get snapshot from this PR, 
we already have some none production use cases that are only blocked by sqldelight right now 
we can help test the changes too 😄

IlyaGulya commented:
> I appreciate the huge amount of enthusiasm around this PR, and I want to say that I'm also very excited about the prospect of using SQLDelight in Wasm projects.
> 
> I also really appreciate the contributions being made here, but this PR is very difficult to review as-is because of its size and because of all of the changes that aren't directly related to enabling Wasm that are being pulled into here.
> 
> My recommendation would be to break this PR up into a few smaller PRs that are more focused and will be easier/quicker to review:
> 
> 1. A PR that enables the `wasmJs()` target in the runtime and extensions
> 2. A PR that adds Wasm support for the web worker driver
> 3. Additional PRs to make other gradle config/code cleanliness changes after the fact
> 
> Thank you!

Sure, I will make those changes it in the next couple of weeks

IlyaGulya commented:
@ahna92 you can always publish this branch to your mavenLocal and take a look 🙂 
Making a snapshot is not that easy so I'll just focus on preparing this PR to be merged.

IlyaGulya commented:
@dellisd first PR is here https://github.com/sqldelight/sqldelight/pull/5531 🙂

dellisd commented:
I'll be closing this PR in favour of the new ones!
